File: backend_python/app/models/model_run.py
from datetime import datetime
from typing import Optional, List, Dict, Any, Literal
from pydantic import BaseModel, Field

class ModelRunModel(BaseModel):
    id: str = Field(alias="_id")  # Custom String primary key (e.g. "RUN-CNN-4821")
    batchId: str  # Ref to InspectionBatch._id
    productId: str  # Ref to Product._id
    inputImageId: Optional[str] = None  # Ref to InspectionImage._id
    modelType: Literal["cnn", "unet", "yolo", "detection", "segmentation"]
    modelVersion: str = "v1.0.0"
    status: Literal["queued", "running", "completed", "failed"] = "queued"
    startedAt: datetime = Field(default_factory=datetime.utcnow)
    completedAt: Optional[datetime] = None
    overallConfidence: Optional[float] = None
    outputImageIds: List[str] = []  # List of InspectionImage._id
    rawOutput: Optional[Dict[str, Any]] = None
    errorMessage: Optional[str] = None
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True

File: backend/server/models/ModelRun.ts
import mongoose, { Schema, Document } from "mongoose";

export interface IModelRun extends Document {
  _id: string; // Primary key (e.g. "RUN-CNN-4821")
  batchId: string;
  productId: string;
  inputImageId?: string | null;
  modelType: "cnn" | "unet" | "yolo" | "detection" | "segmentation";
  modelVersion: string;
  status: "queued" | "running" | "completed" | "failed";
  startedAt: Date;
  completedAt?: Date | null;
  overallConfidence?: number | null;
  outputImageIds: string[];
  rawOutput?: Record<string, any> | null;
  errorMessage?: string | null;
  createdAt: Date;
  updatedAt: Date;
}

const ModelRunSchema = new Schema<IModelRun>(
  {
    _id: { type: String, required: true },
    batchId: { type: String, ref: "InspectionBatch", required: true },
    productId: { type: String, ref: "Product", required: true },
    inputImageId: { type: String, ref: "InspectionImage", default: null },
    modelType: {
      type: String,
      enum: ["cnn", "unet", "yolo", "detection", "segmentation"],
      required: true
    },
    modelVersion: { type: String, default: "v1.0.0" },
    status: {
      type: String,
      enum: ["queued", "running", "completed", "failed"],
      default: "queued"
    },
    startedAt: { type: Date, default: Date.now },
    completedAt: { type: Date, default: null },
    overallConfidence: { type: Number, default: null },
    outputImageIds: [{ type: String, ref: "InspectionImage" }],
    rawOutput: { type: Schema.Types.Mixed, default: null },
    errorMessage: { type: String, default: null }
  },
  {
    timestamps: true,
    _id: false
  }
);

ModelRunSchema.index({ productId: 1, modelType: 1, createdAt: -1 });
ModelRunSchema.index({ batchId: 1, createdAt: -1 });

export const ModelRun =
  mongoose.models.ModelRun ||
  mongoose.model<IModelRun>("ModelRun", ModelRunSchema, "modelRuns");
